_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+You can fast-forward to the next section or start your journey by watching this awesome video by [Nate Barbettini](https://twitter.com/nbarbettini) from [Okta](https://www.okta.com/).
+
+<div class="responsive-video">
+    <iframe width="1164" height="655" src="https://www.youtube.com/embed/996OiexHze0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+</div>
+<br>
+
+## Implicit Flow versus Code Flow + PKCE
+In this example, we will use the authorization code grant flow with Proof Key for Code Exchange ([PKCE](https://tools.ietf.org/html/rfc7636)) to secure the Angular app. 
+It's a very long name for what could be shortened to "code flow + PKCE" which is [more secure than the implicit flow](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-13#section-3.1.2).   
+In fact, the implicit flow was never very secure to begin with. 
+This is well-explained by [Nate Barbettini](https://twitter.com/nbarbettini) and [Aaron Parecki](https://twitter.com/aaronpk
+):
+
+<div class="responsive-video">
+    <iframe width="1164" height="655" src="https://www.youtube.com/embed/CHzERullHe8" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+</div>
+<br>
+
+The implicit flow was the easiest to understand, since it required one step less than the standard code flow:
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/implicit_vs_code.png" width="60%">
+</div>
+
+PKCE is an addition on top of the standard code flow to make it usable for public clients. 
+It is already in use for native and mobile clients. 
+PKCE boils down to this: 
+1. Give `hash` of random value to authorization server when logging in to ask for `code`
+2. Hand over the `random value` to authorization server when exchanging `code` for `access token`
+3. Authorization server returns `access token` after verifying that `hash` belongs to `random value`.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/pkce.png" width="70%">
+</div>
+
+If a fraudster were to intercept our authorization grant (the code), he or she would still not have the `code_verifier`, which is stored in our SPA client. 
+If he/she tries to exchange the stolen authorization grant without this value, the response would not be a token but rather an `HTTP 400`:
+
+```json
+{
+    "error": "invalid_grant",
+    "error_description": "PKCE code verifier not specified"
+}
+```
+
+This is why the code flow + PKCE is more secure than the implicit flow. 
+Even if an attacker manages to obtain the authorization grant, it's worthless without the `code_verifier`.
+
+> Note that the HTTP 400 will only occur when using PKCE. 
+If no PKCE is used, the client should be confidential (requiring credentials to exchange the authorization grant) rather than be public.
+
+So only our Angular client will be able to retrieve the access token in the form of a JSON Web Token.
+
+## JSON Web Token (JWT)
```

Comment:
In this chapter, you can refer to another blogpost of ours, https://ordina-jworks.github.io/microservices/2016/05/01/Using-JWT-Tokens-for-State-Transfer.html